controller/volume_controller.go

```diff
@@ -771,33 +790,101 @@ func (vc *VolumeController) cleanupFailedToScheduledReplicas(v *longhorn.Volume,
 
 func (vc *VolumeController) cleanupExtraHealthyReplicas(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (err error) {
 	healthyCount := vc.getHealthyReplicaCount(rs)
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
 
-	if healthyCount > v.Spec.NumberOfReplicas {
-		for _, r := range rs {
-			// Clean up eviction requested replica
-			if r.Status.EvictionRequested {
-				if err := vc.deleteReplica(r, rs); err != nil {
-					vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
-						EventReasonFailedEviction,
-						"volume %v failed to evict replica %v",
-						v.Name, r.Name)
-					return err
-				}
-				logrus.Debugf("Evicted replica %v", r.Name)
-				return nil
+	var cleaned bool
+	cleaned, err = vc.cleanupEvictionRequestedReplicas(v, rs)
+	if err != nil {
+		return err
+	}
+	if cleaned {
+		return nil
+	}
+
+	cleaned, err = vc.cleanupAutoBalancedReplicas(v, rs)
+	if err != nil {
+		return err
+	}
+	if cleaned {
+		return nil
+	}
+
+	cleaned, err = vc.cleanupDataLocalityReplicas(v, e, rs)
+	if err != nil {
+		return err
+	}
+	if cleaned {
+		return nil
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) cleanupEvictionRequestedReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	for _, r := range rs {
+		if r.Status.EvictionRequested {
+			if err := vc.deleteReplica(r, rs); err != nil {
+				vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
+					EventReasonFailedEviction,
+					"volume %v failed to evict replica %v",
+					v.Name, r.Name)
+				return false, err
 			}
+			logrus.Debugf("Evicted replica %v", r.Name)
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (vc *VolumeController) cleanupAutoBalancedReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	log := getLoggerForVolume(vc.logger, v)
+	log = log.WithField("replicaAutoBalanceType", "delete")
```

Comment:
nit: can be combined no need for an independent assignment.
```suggestion
	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceType", "delete")
```

Reply:
Thanks! fixed.